src/boot/bootctl-install.c

```diff
@@ -449,7 +449,84 @@ static int copy_one_file(const char *esp_path, const char *name, bool force) {
         return ret;
 }
 
-static int install_binaries(const char *esp_path, const char *arch, bool force) {
+static int install_binaries_extra(const char *esp_path, const char *arch, bool force, char** ret_default_loader_path) {
+        char *root = IN_SET(arg_install_source, ARG_INSTALL_SOURCE_AUTO, ARG_INSTALL_SOURCE_IMAGE) ? arg_root : NULL;
+        _cleanup_closedir_ DIR *d = NULL;
+        _cleanup_free_ char *path = NULL;
+        _cleanup_free_ char *default_loader_path = NULL;
+        int r;
+
+        r = chase_and_opendir(BOOTLIBDIR "/extra", root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, &path, &d);
+        /* If we had a root directory to try, we didn't find it and we are in auto mode, retry on the host */
+        if (r == -ENOENT && root && arg_install_source == ARG_INSTALL_SOURCE_AUTO)
+                r = chase_and_opendir(BOOTLIBDIR "/extra", NULL, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, &path, &d);
+        if (r < 0)
+                return r; /* ignore if vendor dir does not exist */
+
+        FOREACH_DIRENT(de, d, continue) {
+                if(!is_dir_at(dirfd(d), de->d_name, true))
+                        continue;
+
+                _cleanup_closedir_ DIR *dd = xopendirat(dirfd(d), de->d_name, 0);
+                if (!dd) {
+                        log_error_errno(errno, "Failed to open %s/%s: %m", path, de->d_name);
+                        continue;
+                }
+
+                FOREACH_DIRENT(dde, dd, continue) {
+                        _cleanup_free_ char *source_path = path_join(path, de->d_name, dde->d_name);
+                        _cleanup_free_ char *dest_path = NULL, *p = NULL;
+                        int k;
+
+                        if (!source_path)
```

Comment:
we generally try to keep function calls and their immmediate error handling together. hence please move the path_join() call closer to this error handling of it, please